experimental/personnames/src/derive_missing_initials.rs

```diff
@@ -0,0 +1,170 @@
+use crate::api::{FieldLength, FieldModifier, NameField, PersonName};
+
+const CLDR_INITIAL_INITIAL_PATTERN: &str = r"\{(?P<initial_position>\d+)}(?P<trailing>[^{]+)?";
+const CLDR_INITIAL_SEQUENCE_PATTERN: &str = r"\{(?P<initial_position>\d+)}(?P<trailing>[^{]+)?";
+
+lazy_static! {
+    static ref INITIAL_PATTERN: Regex = Regex::new(CLDR_INITIAL_INITIAL_PATTERN).unwrap();
+    static ref INITIAL_PATTERN_SEQUENCE: Regex = Regex::new(CLDR_INITIAL_SEQUENCE_PATTERN).unwrap();
+}
+
+///
+/// Derive missing initials from the name
+///
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#derive-initials
+///
+pub(crate) fn derive_missing_initials(
+    person_name: &dyn PersonName,
+    requested_field: &NameField,
+    initial_pattern: &str,
+    initial_sequence_pattern: &str,
+) -> String {
+    if person_name.has_name_field(requested_field) {
+        return person_name.get(requested_field).to_string();
+    }
+    if requested_field.modifier.has_field(FieldModifier::Initial) {
+        let full_value = Vec::from_iter(
+            person_name
+                .get(&NameField {
+                    kind: requested_field.kind,
+                    modifier: requested_field.modifier.with_length(FieldLength::Auto),
+                })
+                .split(' ')
+                .map(|s| s.trim()),
+        );
+        let captures: CaptureMatches = INITIAL_PATTERN.captures_iter(initial_pattern);
+        let captures_sequence: CaptureMatches =
+            INITIAL_PATTERN_SEQUENCE.captures_iter(initial_sequence_pattern);
+
+        let mut initials = vec![];
+        for capture in captures {
+            let value_iter = full_value.iter();
+            let position: usize = capture
+                .name("initial_position")
+                .map(|m| m.as_str().parse().unwrap())
+                .unwrap();
+
+            let trailing = capture.name("trailing").map_or("", |m| m.as_str());
+            for name_part in value_iter {
+                let initial = name_part
+                    .chars()
+                    .nth(position)
+                    .map(|i| i.to_string() + trailing)
+                    // Default to empty string if not found.
+                    .unwrap_or(String::from(""));
+                initials.push(initial);
+            }
+        }
+        let mut build_initials = vec![];
+
+        for capture_sequence in captures_sequence {
+            let position: usize = capture_sequence
+                .name("initial_position")
+                .map(|m| m.as_str().parse().unwrap())
+                .unwrap();
+            let trailing_sequence = capture_sequence.name("trailing").map_or("", |m| m.as_str());
+            build_initials.push(
+                initials
+                    .get(position)
+                    .map(|s| s.to_string() + trailing_sequence)
+                    .unwrap_or(trailing_sequence.to_string()),
+            )
+        }
+        return build_initials.join("").trim().to_string();
```

Comment:
This function seems overly heavy in allocating vec and string operations. Haven't thought through if there's a better way to implement it, but it seems like you should be able to do it with only allocating the return value and not all these intermediate ones.